### DISCUSSION TITLE
Separate client from server (monolithic style)

### DISCUSSION THREAD

**Discussion by eliandoran** [TriliumNext/Notes#12] [Ideas]
From a cursory look at the architecture of the application, it appears that Trilium is implemented as a monolith where both the client and the server reside in the same repository.

This architecture is good since it allows easy iteration of both the client and the server without having to deal with multiple repositories.

However, what I'm seeing is that the Trilium is a single-package monolith: there is a single `package.json` which contains dependencies for both the client and the server (as well as Electron).

One thing I cannot wrap my head around is that the client resides in `src/public/app` whereas the server is in `src`. Both the client and the server have their own `services` directory.

```
src
  public
    app: client
      services
      components
  services
package.json
```

To me the structure should be:

```
client/src:
  services
  components
  package.json
server/src:
  services
  components
  package.json
```

So the goal would be to physically separate the client and the server into separate directories, while maintaining a single repo. I think Trilium is implemented as a single project in order to simplify dependency management but we can benefit from tools such as Turbo to handle the monorepo for us.

What is your opinion on this?

**Comment by eliandoran** [comment]
@zadam , I'd love your opinion on this one.
I'm also wondering if you were to rewrite Trilium now, what would you do differently?

**Reply by zadam** [reply]
In general I prefer monorepo approach, I think it keeps things simple.

Regarding the project structure, backend and frontend are separate apps, they should have separate directory trees, separate package.json.

**Comment by eliandoran** [comment]
Created #21 to track the task.
